File: petclinic-backend-ts/src/pets/dto/pet.dto.ts
import { components } from '../../generated/api-types';
import { Exact } from '../../common/contract';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import { IsDefined, IsOptional, Matches, MaxLength, Min, ValidateNested } from 'class-validator';
import { PetTypeDto } from './pet-type.dto';
import { VisitDto } from '../../visits/dto/visit.dto';
import { IsPastOrPresent } from '../../common/validation/is-past-or-present.validator';

/**
 * Data transfer object for a pet.
 *
 * birthDate is an ISO 'YYYY-MM-DD' string.
 */
export class PetDto {
  // Reject null, empty AND whitespace-only with a single "must not be blank":
  // @IsDefined covers null; @Matches(/\S/) covers empty/whitespace; @MaxLength caps length at 30.
  @IsDefined({ message: 'must not be blank' })
  @Matches(/\S/, { message: 'must not be blank' })
  @MaxLength(30, { message: 'size must be between 0 and 30' })
  @ApiProperty({ example: 'Leo', description: 'The name of the pet.' })
  name!: string;

  @IsDefined({ message: 'must not be null' })
  @IsPastOrPresent({ message: 'Birth date must not be in the future' })
  @ApiProperty({ example: '2010-09-07' })
  birthDate!: string;

  @IsDefined({ message: 'must not be null' })
  @ValidateNested()
  @Type(() => PetTypeDto)
  type!: PetTypeDto;

  @IsOptional()
  @Min(0, { message: 'must be greater than or equal to 0' })
  @ApiProperty({
    readOnly: true,
    example: 1,
    description: 'The ID of the pet.',
    required: true,
  })
  id!: number;

  @IsOptional()
  @Min(0, { message: 'must be greater than or equal to 0' })
  @ApiPropertyOptional({ readOnly: true, example: 1, description: "The ID of the pet's owner." })
  ownerId?: number;

  @ValidateNested({ each: true })
  @Type(() => VisitDto)
  @ApiProperty({
    type: () => [VisitDto],
    readOnly: true,
    description: 'Vet visit bookings for this pet.',
    required: true,
  })
  visits: readonly VisitDto[] = [];
}

// Compile-time lock against the root openapi.yaml (see GUARDRAILS.md).
true satisfies Exact<PetDto, components['schemas']['PetDto']>;

File: petclinic-backend/src/main/java/victor/training/petclinic/rest/dto/PetDto.java
package victor.training.petclinic.rest.dto;

import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.PastOrPresent;
import jakarta.validation.constraints.Size;
import org.springframework.lang.Nullable;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class PetDto {
    @NotBlank
    @Size(max = 30)
    @Schema(example = "Leo", description = "The name of the pet.")
    private String name;

    @NotNull
    @Valid
    @PastOrPresent(message = "Birth date must not be in the future")
    @Schema(example = "2010-09-07")
    private LocalDate birthDate;

    @NotNull
    @Valid
    private PetTypeDto type;

    @Min(0)
    @Schema(accessMode = Schema.AccessMode.READ_ONLY, example = "1", description = "The ID of the pet.",
            requiredMode = Schema.RequiredMode.REQUIRED)
    private Integer id;

    @Min(0)
    @Schema(accessMode = Schema.AccessMode.READ_ONLY, example = "1", description = "The ID of the pet's owner.")
    private @Nullable Integer ownerId;

    @Valid
    @Schema(accessMode = Schema.AccessMode.READ_ONLY, description = "Vet visit bookings for this pet.",
            requiredMode = Schema.RequiredMode.REQUIRED)
    private List<VisitDto> visits = new ArrayList<>();

    public String getName() {
        return name;
    }

    public PetDto setName(String name) {
        this.name = name;
        return this;
    }

    public LocalDate getBirthDate() {
        return birthDate;
    }

    public PetDto setBirthDate(LocalDate birthDate) {
        this.birthDate = birthDate;
        return this;
    }

    public PetTypeDto getType() {
        return type;
    }

    public PetDto setType(PetTypeDto type) {
        this.type = type;
        return this;
    }

    public Integer getId() {
        return id;
    }

    public PetDto setId(Integer id) {
        this.id = id;
        return this;
    }

    public Integer getOwnerId() {
        return ownerId;
    }

    public PetDto setOwnerId(Integer ownerId) {
        this.ownerId = ownerId;
        return this;
    }

    public List<VisitDto> getVisits() {
        return visits;
    }

    public PetDto setVisits(List<VisitDto> visits) {
        this.visits = visits;
        return this;
    }
}
